x-pack/plugins/enterprise_search/public/applications/shared/api_logic/create_api_logic.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { kea, MakeLogicType } from 'kea';
+
+import { ApiStatus, Status, HttpError } from '../../../../common/types/api';
+import { flashAPIErrors, clearFlashMessages } from '../flash_messages';
+
+export interface Values<T> {
+  apiStatus: ApiStatus<T>;
+  status: Status;
+  data?: T;
+  error: HttpError;
+}
+
+export interface Actions<Args extends Record<string, unknown> | undefined, Result> {
+  initiateCall(args: Args): Args;
+  apiError(code: number, error: string): HttpError;
+  apiSuccess(result: Result): Result;
+  apiReset(): void;
+}
+
+export const createApiLogic = <Result, Args extends Record<string, unknown> | undefined>(
+  path: string[],
+  apiFunction: (args: Args) => Promise<Result>
+) =>
+  kea<MakeLogicType<Values<Result>, Actions<Args, Result>>>({
+    path: ['enterprise_search', 'workplace_search', ...path],
+    actions: {
+      initiateCall: (args) => args,
+      apiError: (code, error) => ({ code, error }),
+      apiSuccess: (result) => result,
+      apiReset: true,
+    },
+    reducers: () => ({
+      apiStatus: [
```

Comment:
It seems strange to me to clump this together in a single reducer, and then use selectors to split them apart again.  We could still re-enforce the logic you're describing at the reducer level

```
export interface Values<T> {
  status: Status;
  data: null | T;
  error: null | HttpError;
}
```

```
  reducers: () => ({
      status: [
        Status.IDLE,  
        {
          makeRequest: () => Status.LOADING,
          apiError: () => Status.ERROR,
          apiSuccess: () => Status.SUCCESS,
        },
      ],
      data: [
        null,
        {
          makeRequest: () => null,
          apiSuccess: (_, data) => data,
        }
      ],
      error: [
        null,
        {
          makeRequest: () => null,
          apiError: (_, error) => error,
        }
      ]
    }),
```

Reply:
I tried this, and then ran into the issue where you can't set a value to 'undefined' because...eh....Kea is weird. Annoying. I'm not a fan of using `null` there tbh.

Reply:
The null/undefined thing is a reality of using kea, we use the null pattern a lot (and I prefer it much more than setting things as empty objects).  I still think composing all the values into effectively `Logic.values.values` is not an ideal pattern, but it won't change the functionality of this so i'm fine with it.